03-rec-poly/homework-03/hw-03.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-
+Exercise 1 Hopscotch
+Your first task is to write a function
+cis 194: homework 3 2
+skips :: [a] -> [[a]]
+The output of skips is a list of lists. The first list in the output should be the same as the input list. The second list in the output should contain every second element from the input list. . . and the nth list in the output should contain every nth element from the input list.
+For example:
+== ["ABCD", "BD", "C", "D"]
+== ["hello!", "el!", "l!", "l", "o", "!"]
+== [[1]]
+skips "ABCD"
+skips "hello!"
+skips [1]
+skips [True,False] == [[True,False], [False]]
+skips []           == []
+Note that the output should be the same length as the input.
+-}
+--Exercise 1
+skips :: forall a. [a] -> [[a]]
+skips xs = map (\(idx, _) -> filterIdx idx ps) ps
+  where
+    ps :: [(Int, a)]
+    ps = zip [1 .. length xs] xs
+    filterIdx :: Int -> [(Int, a)] -> [a]
+    filterIdx n ls = map snd $ filter (\(idx, _) -> idx `mod` n == 0) ls
+
+
+{-
+Exercise 2 Local maxima
+A local maximum of a list is an element of the list which is strictly
+greater than both the elements immediately before and after it. For example, in the list [2,3,4,1,5], the only local maximum is 4, since it is greater than the elements immediately before and after it (3 and 1). 5 is not a local maximum since there is no element that comes after it.
+Write a function
+localMaxima :: [Integer] -> [Integer]
+which finds all the local maxima in the input list and returns them in order. For example:
+ localMaxima [2,9,5,6,1] == [9,6]
+ localMaxima [2,3,4,1,5] == [4]
+ localMaxima [1,2,3,4,5] == []
+-}
+--Exercise 2
+localMaxima :: [Integer] -> [Integer]
+localMaxima (x : y : z : zs)
+    | y > x && z > y = y : localMaxima (z : zs)
+    | otherwise      = localMaxima (y : z : zs)
```

Comment:
What happens in case the list is empty, or has 1 or 2 elements?

Reply:
@oishikg It should return an empty list, this one is my oversight, I didn't went back to check Exercise 2 after I have moved on (other pattern match deleted from undoing)